drivers/i2c/i2c_ite_enhance.c

```diff
@@ -568,6 +605,311 @@ static int i2c_enhance_transfer(const struct device *dev, struct i2c_msg *msgs,
+			enhanced_i2c_cq_read(dev);
+		/* I2C write of command queue mode. */
+		} else {
+			enhanced_i2c_cq_write(dev);
+		}
+	}
+
+	/* Enable i2c module with command queue mode. */
+	IT8XXX2_I2C_CTR1(base) = IT8XXX2_I2C_MDL_EN | IT8XXX2_I2C_COMQ_EN;
+	/* One shot on device 1. */
+	IT8XXX2_I2C_MODE_SEL(base) = 0;
+	IT8XXX2_I2C_CTR2(base) = 1;
+	/* Start */
+	i2c_enhance_disable_cpu_idle(config->port);
+	IT8XXX2_I2C_CTR(base) = E_START_CQ;
+
+	return 1;
+}
+
+static int i2c_enhance_cq_transfer(const struct device *dev,
+				   struct i2c_msg *msgs)
+{
+	struct i2c_enhance_data *data = dev->data;
+	const struct i2c_enhance_config *config = dev->config;
+	int res = 0;
+
+	data->err = 0;
+	data->cq_msgs = msgs;
+
+	/* Make sure we're in a good state to start */
+	if (i2c_bus_not_available(dev)) {
+		/* Recovery I2C bus */
+		i2c_recover_bus(dev);
+		/*
+		 * After resetting I2C bus, if I2C bus is not available
+		 * (No external pull-up), drop the transaction.
+		 */
+		if (i2c_bus_not_available(dev)) {
+			return -EIO;
+		}
+	}
+
+	/* Start transaction */
+	if (enhanced_i2c_cmd_queue_trans(dev)) {
+		/* Enable i2c interrupt */
+		irq_enable(config->i2c_irq_base);
+	}
+	/* Wait for the transfer to complete */
+	res = k_sem_take(&data->device_sync_sem, K_MSEC(100));
+
+	irq_disable(config->i2c_irq_base);
+
+	if (res != 0) {
+		data->err = ETIMEDOUT;
+		/* Reset i2c port. */
+		i2c_reset(dev);
+		LOG_ERR("I2C ch%d:0x%X reset cause %d",
+			config->port, data->addr_16bit, I2C_RC_TIMEOUT);
+	}
+
+	i2c_enhance_enable_cpu_idle(config->port);
+
+	return data->err;
+}
+
+static int cq_mode_allowed(const struct device *dev, struct i2c_msg *msgs)
+{
+	struct i2c_enhance_data *data = dev->data;
+
+	/*
+	 * If there is only one message can be got, we can do command queue
+	 * transfer directly.
+	 * If the length of write or read transfer is greater than command
+	 * queue payload size, there will execute PIO mode.
+	 */
+	if (data->num_msgs == 1 && (msgs[0].flags & I2C_MSG_STOP)) {
+		/* e.g. msg.flags = I2C_MSG_READ | I2C_MSG_STOP; */
+		if ((msgs[0].flags & I2C_MSG_READ) &&
+		    (msgs[0].len <= CONFIG_I2C_CQ_MODE_MAX_PAYLOAD_SIZE)) {
+			return 1;
+		/*
+		 * e.g. msg.flags = I2C_MSG_WRITE | I2C_MSG_STOP;
+		 *
+		 * Note: The cmd_i2c_scan function in the i2c_shell.c only
+		 *       writes the address without writing any data (just
+		 *       confirms whether the device exists).
+		 *       The command queue mode is not applicable here, so
+		 *       the PIO mode is used.
+		 */
+		} else if (((msgs[0].flags & I2C_MSG_RW_MASK) == I2C_MSG_WRITE) &&
+			   (msgs[0].len <= I2C_CQ_MODE_TX_MAX_PAYLOAD_SIZE) &&
+			   (msgs[0].len != 0)) {
+			return 1;
+		} else {
+			return 0;
+		}
+	/*
+	 * If there are two messages, we need to judge whether or not there is
+	 * I2C_MSG_RESTART flag from the second message, and then decide to do
+	 * the command queue or PIO mode transfer.
+	 */
+	} else if (data->num_msgs == 2) {
```

Comment:
No need for the else - the previous if block always returns.

Reply:
Done.

Reply:
Done.